### DISCUSSION TITLE
[2021/09/14] Frontend Night

### DISCUSSION THREAD

**Discussion by jjenko** [wantedly/frontend-night#163] [General]
## Policy

- 資料の準備をがんばらない
- 参加・途中退席は自由
- スキップしない

話したい・聞きたいネタを書いてく ✏️

ハッシュタグ [#frontend_night](https://twitter.com/hashtag/frontend_night?src=hash)

Prev: https://github.com/wantedly/frontend-night/issues/160 

 _Created from https://github.com/wantedly/frontend-night/issues/65 by [issue-creator](https://github.com/rerost/issue-creator)_

**Comment by kobayang** [comment]
## [GraphQLとクライアントサイドの実装指針](https://gist.github.com/Quramy/566ea87d0121ceb8cd97ad9d14b63fd8)

- Demand Driven Architecture(DDA)とは、データを取得する側が、どのようなレスポンスを必要とするかを宣言する方式
- ComponentとFragmentをセットで配置することをFragment Colocation と呼ぶ

📝  Component の 木構造と Fragment の依存関係は同じになる

**Comment by izumin5210** [comment]
## [TypedDocumentNode: the next generation of GraphQL and TypeScript \- The Guild Blog](https://the-guild.dev/blog/typed-document-node)

- pre-compile された GraphQL Operation（`DocumentNode`, `` gql`...` `` の返り値）に TypeScript で型を当てたもの
- (client library 側が対応していれば) アプリ側のコードは型を指定しなくても、Operation から型を推論してくれるようになる

かなり良さそうだったので、おもむろに突っ込んでみている https://github.com/wantedly/wantedly-frontend/pull/3345

**Comment by kobayang** [comment]
[フロントエンドテストプラクティス in open 8](https://speakerdeck.com/yosuke_furukawa/hurontoendotesutopurakuteisu-in-open-8)

- テストとは?
  - 品質を上げるもの
  - 将来の変更を容易にしてくれるもの
  - 不具合を見つけてくれるための未来への投資
- test = check + explore
  - 理想のテストは確認(Check)と探索(Explore)の両方しているテスト
  - 開発者、QAの両方が check の状態では test にならない。探索して発見するプロセスを入れる必要がある。
- フロントエンドのテストのベストプラクティス

![スクリーンショット 2021-09-14 10 53 28](https://user-images.githubusercontent.com/12181179/133181654-e7a36be9-b9dc-4380-8978-c12fcf8e040a.png)

wantedly の構成で考えると...

- unit test → src/components 以下に書く
- integration test → src/pages 以下に書く
- e2e → /cypress に書く
- a11y / Performance → ???

合わせて読む: [フロントエンドのテストは皆のためのもの](https://postd.cc/frontend-testing-is-for-everyone/)

**Reply by kobayang** [reply]
- 振る舞いのうちよりコアな体験をe2eで守る。
- エッジケースは integration

**Reply by kobayang** [reply]
参考

https://kentcdodds.com/blog/static-vs-unit-vs-integration-vs-e2e-tests
https://twitter.com/kentcdodds/status/960723172591992832
